### ISSUE TITLE
google.location == us-east5 in Nextflow config leads to wrong machine type choices

### ISSUE THREAD

**Issue by aaronegolden** [nextflow-io/nextflow#5602]
cloudinfo.seqera.io seems to be missing information for us-east5

https://cloudinfo.seqera.io/api/v1/providers/google/services/compute/regions/us-east5/products

yields

```
{
  "type": "about:blank",
  "title": "validation problem",
  "status": 400,
  "detail": "Key: 'GetRegionPathParams.Region' Error:Field validation for 'Region' failed on the 'region' tag"
}
```

and as a consequence the GoogleBatchMachineTypeSelector can't find any available machine types for us-east5 here:

https://github.com/nextflow-io/nextflow/blob/master/plugins/nf-google/src/main/nextflow/cloud/google/batch/GoogleBatchMachineTypeSelector.groovy#L144

which can lead to NextFlow submitting Batch jobs with incorrect machine types.

**Comment by pditommaso** [comment]
Is it reporting an error? can you please include the `.nextflow.log` file?

**Comment by tony-vs** [comment]
When comparing the returned data for the us-east5 region to the GCP Compute product listing for the us-east1 region (e.g., [us-east1 region data](https://cloudinfo.seqera.io/api/v1/providers/google/services/compute/regions/us-east1/products)), it becomes evident that the us-east5 data contains invalid information.

**Comment by jorgee** [comment]
In this case, the job should fail instead of submitting it, is it correct?

**Comment by pditommaso** [comment]
Likely it should report a warning and fallback on the default mechanism that's this 

https://github.com/nextflow-io/nextflow/blob/6183fdf9dbd9114f5ee86d11bc5e69cbd06501a6/plugins/nf-google/src/main/nextflow/cloud/google/batch/GoogleBatchMachineTypeSelector.groovy#L107-L107

**Comment by tony-vs** [comment]
Are we expecting a valid response from the API call? The us-east5 region is valid, so the data returned should reflect this.

> In this case, the job should fail instead of submitting it, is it correct?

**Comment by pditommaso** [comment]
It should but, for some reason that we are investigating, it's failing. In any case, Nextflow should be resilient to these error conditions.

**Comment by jorgee** [comment]
I have run a pipeline using us-east5 and I see there is already a fallback. Looking at the .nextflow.log, I can see the cloudInfo call returns error 400. It generates an exception that is catched and the job has been scheduled and running with an instance type selected by Google Batch in the selected region.

The log file show the following:

```
Dec-17 20:47:02.215 [Task submitter] DEBUG n.c.g.batch.GoogleBatchTaskHandler - [GOOGLE BATCH] Cannot select machine type using Seqera Cloudinfo for task: `TEST (1)` - Server returned HTTP response code: 400 for URL: https://cloudinfo.seqera.io/api/v1/providers/google/services/compute/regions/us-east5/products
Dec-17 20:47:02.217 [Task submitter] TRACE n.c.g.batch.GoogleBatchTaskHandler - [GOOGLE BATCH] new job request > task_groups {
  task_spec {
    compute_resource {
      cpu_milli: 1000
    }
    volumes {
      gcs {
        remote_path: "jorgee-test"
      }
      mount_path: "/mnt/disks/jorgee-test"
      mount_options: "-o rw"
      mount_options: "-implicit-dirs"
    }
    runnables {
      container {
        image_uri: "nextflow/rnaseq-nf"
        commands: "/bin/bash"
        commands: "-o"
        commands: "pipefail"
        commands: "-c"
        commands: "trap \"{ cp .command.log /mnt/disks/jorgee-test/work/e0/b3bbe04c961de9304e1623bfd95638/.command.log; }\" ERR; /bin/bash /mnt/disks/jorgee-test/work/e0/b3bbe04c961de9304e1623bfd95638/.command.run 2>&1 | tee .command.log"
        volumes: "/mnt/disks/jorgee-test/work/e0/b3bbe04c961de9304e1623bfd95638:/mnt/disks/jorgee-test/work/e0/b3bbe04c961de9304e1623bfd95638:rw"
      }
      environment {
      }
    }
  }
}
allocation_policy {
  instances {
    policy {
      provisioning_model: SPOT
    }
  }
}
logs_policy {
  destination: CLOUD_LOGGING
}

Dec-17 20:47:02.866 [Task submitter] DEBUG n.c.g.batch.GoogleBatchTaskHandler - [GOOGLE BATCH] Process `TEST (1)` submitted > job=nf-e0b3bbe0-1734464817649; uid=nf-e0b3bbe0-173446-e5a1ec47-c3a6-44160; work-dir=gs://jorgee-test/work/e0/b3bbe04c961de9304e1623bfd95638
Dec-17 20:47:02.867 [Task submitter] INFO  nextflow.Session - [e0/b3bbe0] Submitted process > TEST (1)
```

nnd the Google Batch service dashboard I can see the selected instance_type for the job:
![image](https://github.com/user-attachments/assets/156d276d-a8ff-4947-9530-fb6fb478b2e4)


In [this part of the code](https://github.com/nextflow-io/nextflow/blob/fc2ac7d1d747d6d9b28c8d6a977b9d3ad215c4ed/plugins/nf-google/src/main/nextflow/cloud/google/batch/GoogleBatchTaskHandler.groovy#L594), Nextflow invoke the [GoogleBatchMachineTypeSelector.bestMachineType](https://github.com/nextflow-io/nextflow/blob/fc2ac7d1d747d6d9b28c8d6a977b9d3ad215c4ed/plugins/nf-google/src/main/nextflow/cloud/google/batch/GoogleBatchMachineTypeSelector.groovy#L92) method that calls the CloudInfo service in [GoogleBatchMachineTypeSelector.getAvailableMachineTypes](https://github.com/nextflow-io/nextflow/blob/fc2ac7d1d747d6d9b28c8d6a977b9d3ad215c4ed/plugins/nf-google/src/main/nextflow/cloud/google/batch/GoogleBatchMachineTypeSelector.groovy#L144C33-L144C57) throwing the exception when error 400 is produced. The exception is catch return null to do not include the instance type in the Jobs's allocation policy. When no instance type is provided Google Batch automatically decides which instance to use.

From the Nextflow point of view, I would just add the warning message to notify this situation to the user. I think there is no need to modify the fallback mechanism.

**Comment by tony-vs** [comment]
If a user specifies a particular machineType to use, the fallback machine will not align with their request. Should the process fail in such cases to ensure accurate provisioning?

**Comment by tony-vs** [comment]
You're correct. Currently, for any machineType with a pattern like n1-* that includes GPU requirements, it always falls back to e2-* machines, which do not satisfy the user's GPU requests.

> If the user specifies a particular machineType, then there is no fallback. the machineType is it is set to the BatchJob. However, the case where I can see an issue is when the user specifies a family with a pattern such as 'n1-*'. Maybe it is the case that @tony-vs is referring to. In this case, Google Batch can not fulfill the family directive. I can't see any way to specify an instance family/series in the Job request description, so I think Nextflow should raise the failure in this case.

**Comment by pditommaso** [comment]
> However, the case where I can see an issue is when the user specifies a family with a pattern such as 'n1-*'.

@jorgee can you provide more details about this?

**Comment by jorgee** [comment]
[Here](https://github.com/nextflow-io/nextflow/blob/1fd5dc5e3535160d688e59ffe64e97922d888d88/plugins/nf-google/src/main/nextflow/cloud/google/batch/GoogleBatchTaskHandler.groovy#L613), it is checking if the MachineType is a particular machineType, a list (separated by commas) or a family (with *). It only sets the MachineType if the user provides a particular type. When there is a list, we could choose one of them (for instance the first one) but when the user provides a family, we can not set any hint to Google Batch because the API only support to set a particular machine type in the allocation policy. So the selected instance by Google Batch could not match with the family specified by the user. 

There is an Google Compute API call to get the types, that we could use but I think we will re-implement the CloudInfo functionality. Therefore, I think the best in this case is to raise a failure to the user suggesting to specify a single instance.

**Comment by pditommaso** [comment]
My point is that if a concrete `machineType` is provided (no list, not `*`) it should retained if cloudinfo is disabled or fails, because it will take this branch 

https://github.com/nextflow-io/nextflow/blob/cd410916051f95c3f199504ad2fb42feb316edba/plugins/nf-google/src/main/nextflow/cloud/google/batch/GoogleBatchTaskHandler.groovy#L614-L618

and here 

https://github.com/nextflow-io/nextflow/blob/cd410916051f95c3f199504ad2fb42feb316edba/plugins/nf-google/src/main/nextflow/cloud/google/batch/GoogleBatchTaskHandler.groovy#L356-L356

Is that correct?

**Comment by jorgee** [comment]
Yes, it is correct in that case there is no issue. The issue is that if it provides a family. It will continue without any machine type in the instance policy, so the selected instance could not fit with the user directive and could produce a failure. We could warn and continue (assuming that the job could fail) or raise a failure without submitting the job.

**Comment by pditommaso** [comment]
I would limit to report a warning when Cloud info fails to fetch the required metadata. 

Regarding [this](https://github.com/nextflow-io/nextflow/issues/5602#issuecomment-2551795627) comment, the problem looks related when the it's specified a machine type and a GPU. 

However I was wonder if the problem is only when cloud info is failing or in any case because I don't see any condition in the logic below related to GPU filtering 

https://github.com/nextflow-io/nextflow/blob/cd410916051f95c3f199504ad2fb42feb316edba/plugins/nf-google/src/main/nextflow/cloud/google/batch/GoogleBatchMachineTypeSelector.groovy#L93-L128
